packages/gatsby-core-utils/src/service-lock.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Service lock: handles service discovery for Gatsby develop processes
+ * The problem:  the develop process starts a proxy server, the actual develop process and a websocket server for communication. The two latter ones have random ports that need to be discovered. We also cannot run multiple of the same site at the same time.
+ * The solution: lockfiles! We create a lockfolder in `.config/gatsby/sites/${sitePathHash} and then write a file to that lockfolder for every service with its port.
+ *
+ * NOTE(@mxstbr): This is NOT EXPORTED from the main index.ts due to this relying on Node.js-specific APIs but core-utils also being used in browser environments. See https://github.com/jprichardson/node-fs-extra/issues/743
+ */
+import path from "path"
+import tmp from "tmp"
+import lockfile from "proper-lockfile"
+import fs from "fs-extra"
+import xdgBasedir from "xdg-basedir"
+import { createContentDigest } from "./create-content-digest"
+import { isCI } from "./ci"
+
+const globalConfigPath = xdgBasedir.config || tmp.fileSync().name
+
+const getLockfileDir = (programPath: string): string => {
+  // NOTE(@mxstbr): Unsure as to the root cause, but sometimes the createContentDigest
+  // fn adds whitespace at the beginning of the string.
+  const hash = createContentDigest(programPath).trim()
+
+  return path.join(globalConfigPath, `gatsby`, `sites`, `${hash}.lock`)
```

Comment:
The folder is the actual lock"file" that gets checked! The reason the `proper-lockfile` uses "lockdirectories" instead of lock"files" is explained in their readme:

> This library utilizes the mkdir strategy which works atomically on any kind of file system, even network based ones.

While we don't actually care about that specific feature, we do care about others of the library so we go with the lockdirectory anyway!